UnitsNet/CustomCode/UnitAbbreviationsCache.cs

```diff
@@ -247,128 +309,147 @@ public string[] GetUnitAbbreviations(Type unitType, int unitValue, IFormatProvid
 
-            foreach(var enumValue in enumValues)
+            if (formatProvider is not CultureInfo culture)
             {
-                if(TryGetUnitAbbreviations(enumValue, formatProvider, out var abbreviations))
-                {
-                    foreach(var abbrev in abbreviations)
-                    {
-                        unitAbbreviationsPairs.Add((enumValue, abbrev));
-                    }
-                }
+                culture = CultureInfo.CurrentCulture;
             }
 
-            return unitAbbreviationsPairs;
+            return GetAbbreviationsWithFallbackCulture(unitInfo, culture);
         }
 
         /// <summary>
         ///    Get all abbreviations for the given unit and culture.
         /// </summary>
         /// <param name="unitInfo">The unit.</param>
-        /// <param name="formatProvider">The culture to get localized abbreviations for. Defaults to <see cref="CultureInfo.CurrentCulture"/>.</param>
+        /// <param name="culture">The culture to get localized abbreviations for.</param>
         /// <returns>The list of abbreviations mapped for this unit. The first in the list is the primary abbreviation used by ToString().</returns>
         /// <exception cref="ArgumentNullException"><paramref name="unitInfo"/> was null.</exception>
-        public IReadOnlyList<string> GetAbbreviations(UnitInfo unitInfo, IFormatProvider? formatProvider = null)
+        internal IReadOnlyList<string> GetAbbreviationsWithFallbackCulture(UnitInfo unitInfo, CultureInfo culture) 
         {
-            if (unitInfo == null) throw new ArgumentNullException(nameof(unitInfo));
-            if (formatProvider is not CultureInfo)
-                formatProvider = CultureInfo.CurrentCulture;
+            IReadOnlyList<string> abbreviations = GetAbbreviationsForCulture(unitInfo, culture);
 
-            var culture = (CultureInfo)formatProvider;
-            var cultureName = GetCultureNameOrEnglish(culture);
-
-            AbbreviationMapKey key = GetAbbreviationMapKey(unitInfo, cultureName);
-
-            IReadOnlyList<string> abbreviations = AbbreviationsMap.GetOrAdd(key,
-                valueFactory: _ => ReadAbbreviationsFromResourceFile(unitInfo.QuantityName, unitInfo.PluralName, culture));
-
-            return abbreviations.Count == 0 && !culture.Equals(FallbackCulture)
-                ? GetAbbreviations(unitInfo, FallbackCulture)
+            return abbreviations.Count == 0 && HasFallbackCulture(culture)
+                ? GetAbbreviationsForCulture(unitInfo, FallbackCulture)
                 : abbreviations;
         }
 
+        internal static bool HasFallbackCulture(CultureInfo culture)
+        {
+            // accounting for the fact that we're using the same abbreviations for both "en-US" and the "Invariant" culture (Name == string.Empty) 
+            return culture.Name != string.Empty && culture.Name != FallbackCulture.Name;
```

Comment:
Nit, `!Equals(culture, CultureInfo.InvariantCulture)` is more explicit.

The `culture.Name != FallbackCulture.Name` must stay though, since the name is what matters and not the culture info configuration.

Reply:
Sorry, forgot about this one: `culture.Equals` is a bit more expensive than simply checking the name:
``` csharp
  public override bool Equals([NotNullWhen(true)] object? value)
  {
    if (this == value)
      return true;
    return value is CultureInfo cultureInfo && this.Name.Equals(cultureInfo.Name) && this.CompareInfo.Equals((object) cultureInfo.CompareInfo);
  }
```

and I think there is a potential risk for key-duplication if the user uses a custom culture (derived from `CultureInfo.InvariantCulture`).

As for the constant- I tried it, but it looked weird without the accompanying comment (which got even more vague when the string value got hidden by the constant name 😄 ).

PS This looks weird to me too, but I don't see how to improve it..